Standardized Bellman-Ford function calls

NvanAdrichem opened the pull request:
I found that the bellman_ford function call and return parameters are different in design from Dijkstra's and the BFS in unweighted network. So I have standardized these functions to match said function calls and return parameters.

To do so, I have added all the _path() and _path_length() and single_source_ and all_pairs functions accordingly to that "standard". These all now call a private function _bellman_ford() that performs the actual magic. Furthermore, I added multiple predecessor values when multiple shortest paths exist, just like the Dijkstra function does. Finally, I deprecated (not yet removed in case somebody uses it) the function bellman_ford() with a warning and replaced it with bellman_ford_predecessor_and_distance(), since that is what it actually does.

I already updated all dependent functions, existing tests and doctests. However, before merging additional work regarding unit tests on  the new functions, updating of the documentation and maybe a change of the shortest-path function also checking for negative weights needs to occur. I am willing to do that, but would first like to get your first impression before finishing that part.

dschult commented:
This generally looks good. Thanks!
The travis CI errors look related to updating calls to bellman_ford within the package. I hope those are due to rearranging of arguments. I think its ready to move forward with the items you suggest.

I would prefer not to check for negative weights in shortest_path. But we should add a note in the doc_string that negative weights require special methods and point them to bellman_ford and maybe to johnson (which then uses bellman_ford).

We are also discussing a change in the API for these weighted algorithms to make them return iterators instead of dicts. But you probably shouldn't worry about that and get this in place before that work starts. However that is decided we'll just do the same for bellman_ford as we do for Dijkstra.

NvanAdrichem commented:
Cool, I appreciate that you like it. I will continue with the suggested items, but leave shortest_paths alone for now.

NvanAdrichem commented:
I added the documentation and tests.

While using the additions for another project I am currently working on (disjoint paths), I found that using an empty predecessor list (pred=[]) for source nodes does not quite work for nodes that have zero-weighted loops. Although the distance does not change (still 0), the Bellman-Ford function adds a predecessor to the list as it found another path of the same weight. However, now the predecessor only has 1 one item, although 2 paths should have been found (staying at the node, or using the zero-weighted detour). To solve this, I added an initial predecessor list of [None]  for the source nodes.

I expect the same problem exists with the implementation of Dijkstra's algorithm, if you like this solution I will investigate that and if applicable also change Dijkstra's algorithm. If not, let me remove this and just add a check whether a zero-weight self-loop is in place before you merge.

The implementation of the Floyd-Warshall and Johnson's algorithm currently only keep the first found minimum-weight predecessor, so this problem does not (yet) arise there.

dschult commented:
Your solution for initial values for predecessor when we are tracking multiple shortest paths seems good to me. Why don't you look into dijkstra for similar issues in a separate PR though. I think this one just needs an update to remove the conflicts for a merge. hopefully not much.

NvanAdrichem commented:
Thank you for your feedback. I fixed the merging problems, though it took a bit longer than I hoped for (almost all day, for some reason git gave terrible results merging and it was of course a while since I coded this).

I would appreciate a speedy merge since I have some additional work on disjoint paths that I want to push that depends on this PR, furthermore I would dislike additional work on fixing merge conflicts for this PR.

dschult commented:
Thanks very much!   I know all too well how merging messes up subtle things. I appreciate your effort.

NvanAdrichem commented:
Awesome, thanks!
